File: 03/exercise4-objc/exercise4.m
#import <Foundation/Foundation.h>

struct individual {
	NSString *name;
	int age;
	struct individual *next;
};

void printNamelist(struct individual HEAD) {
	struct individual *cur = &HEAD;
	while (cur != NULL) {
		printf("%s (%d)\n", [cur->name UTF8String], cur->age);
		cur = cur->next;
	}
}

NSString* inputCStringToObjC() {
	NSString *output;
	char* str = (char*) malloc(100);
	scanf("%s", str);
	output = [NSString stringWithCString:str encoding:NSUTF8StringEncoding];
	free(str);
	return output;
}

int main() {
	@autoreleasepool{
	int amount = 0;
	printf("How many persons in this group? ");
	scanf("%d", &amount);
	struct individual namelist[amount];
	for(int i = 0; i < amount; i++) {
		printf("Name of No. %d: ", i+1);
		namelist[i].name = inputCStringToObjC();
		printf("Age of No. %d: ", i+1);
		scanf("%d", &namelist[i].age);
	}
	printf("Data input finished.\n");
	for(int i = 0; i < amount; i++) {
		if(i!=amount-1) {
			namelist[i].next = &(namelist[i+1]);
		} else {
			namelist[i].next = NULL;
		}
	}
	printNamelist(namelist[0]);
	return 0;
	}
}

File: 03/exercise4.cpp
#include <iostream>

using namespace std;

struct individual {
	string name;
	int age;
	individual *next;
};

void printNamelist(individual &HEAD) {
	individual *cur = &HEAD;
	while (cur != NULL) {
		cout << cur->name << " (" << cur->age << ")" << endl;
		cur = cur->next;
	}
}

int main() {
	int amount = 0;
	cout << "How many persons in this group? ";
	cin >> amount;
	individual *namelist = new individual[amount];
	for(int i = 0; i < amount; i++) {
		cout << "Name for No. " << i+1 << ": ";
		cin >> namelist[i].name;
		cout << "Age of No. " << i+1 << ": ";
		cin >> namelist[i].age;
	}
	cout << "Data input finished." << endl;
	for(int i = 0; i < amount; i++) {
		if(i!=amount-1) {
			namelist[i].next = &(namelist[i+1]);
		} else {
			namelist[i].next = NULL;
		}
	}
	printNamelist(namelist[0]);
	delete[] namelist;
	return 0;
}

